pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/TokenUtils.java

```diff
@@ -55,35 +55,43 @@ public static <T extends GenericToken<T>> T nthFollower(T token, int n) {
      */
     // test only
     public static <T extends GenericToken<T>> T nthPrevious(T startHint, T anchor, int n) {
-        if (startHint.compareTo(anchor) >= 0) {
-            throw new IllegalStateException("Wrong left hint, possibly not left enough");
-        }
-        if (n <= 0) {
-            throw new IllegalArgumentException("Offset can't be less than 1");
-        }
+        validateInput(startHint, anchor, n);
+    
         int numAway = 0;
         T target = startHint;
         T current = startHint;
+    
         while (current != null && !current.equals(anchor)) {
             current = current.getNext();
-            // wait "n" iterations before starting to advance the target
-            // then advance "target" at the same rate as "current", but
-            // "n" tokens to the left
             if (numAway == n) {
                 target = target.getNext();
             } else {
                 numAway++;
             }
         }
+    
+        validateOutput(current, anchor, numAway, n);
+    
+        return target;
+    }
+    
+    private static <T extends GenericToken<T>> void validateInput(T startHint, T anchor, int n) {
+        if (startHint.compareTo(anchor) >= 0) {
+            throw new IllegalStateException("Wrong left hint, possibly not left enough");
+        }
+        if (n <= 0) {
+            throw new IllegalArgumentException("Offset can't be less than 1");
+        }
+    }
+    
+    private static <T extends GenericToken<T>> void validateOutput(T current, T anchor, int numAway, int n) {
         if (!Objects.equals(current, anchor)) {
             throw new IllegalStateException("Wrong left hint, possibly not left enough");
-        } else if (numAway != n) {
-            // We're not "n" tokens away from the anchor
+        }
+        if (numAway != n) {
             throw new NoSuchElementException("No such token");
```

Comment:
Store exception value in constant of type: private static and call that constant here